### PR TITLE
feat: introduce UnwrapOpaque type

### LINKED ISSUE #165: Add Transparent to reverse Opaque
Opaque nominal types are incredibly useful, but there are times where you need to un-opaque a type ignorer to use it (e.g. as a key in an object).

I propose the addition of
```ts
export type Transparent<T> = T extends Opaque<infer O, any> ? O : T
```

fails now:
```ts
type MyId = Opaque<string, "MyId">
type Nums = Record<MyId, number>
```
> Element implicitly has an 'any' type because expression of type 'Opaque<string, "MyId">' can't be used to index type 'Record<Opaque<string, "MyId">, number>'

but this would fix that error without completely making Opaque types useless or overly compromising their contract:
```ts
type MyId = Opaque<string, "MyId">
type Nums = Record<Transparent<MyId>, number>
```

### PR DESCRIPTION
**Pull request by jmike**
This PR introduces a new `UnwrapOpaque ` type to remove the `[tag]` from Opaque values.

Why is this necessary?

1. Use an _opaque_ type as object keys
2. Prevent stupid `TS4058` error, i.e. _Return type of exported function has or is using name X from external module Y but cannot be named_.

Fixes #165

### CODE REVIEW

**Comment by sindresorhus** [comment]
Any thoughts on the discussion in https://github.com/sindresorhus/type-fest/issues/165?

**Comment by jmike** [comment]
As far as naming goes, I do not have a preference. I can amend this PR to anything you decide.

Re: `protected`, I don't like classes but they seem to do the trick here. I cannot say - needs extensive testing.

I just thought this discussion is ongoing for quite some time, so why not do something about it since I am also affected by the same problems at work.

**Comment by sindresorhus** [comment]
I'm fine with adding this, but it should be called `UnwrapOpaque`.

**Comment by jmike** [comment]
I have renamed the new type (as well as the PR title and description) to `UnwrapOpaque` per @sindresorhus suggestion.

File: source/opaque.d.ts [hunk 1]
```diff
@@ -72,3 +72,36 @@ type Person = {
 @category Type
 */
 export type Opaque<Type, Token = unknown> = Type & Tagged<Token>;
+
+/**
+Remove readonly [tag] from an Opaque type, essentially reverting it back to a generic (i.e. non-Opaque) type.
```
Comment (sindresorhus) [inline source/opaque.d.ts:77]: I would start this sentence with **what** it does, and then explain in another sentence how it does it.

> Reverts an `Opaque` type back to its original type. [...]

**Comment by sindresorhus** [comment]
Looks good.

You need to add it to the readme.

File: source/opaque.d.ts (hunk 1 above)
Comment (jmike) [reply source/opaque.d.ts]: Done.

**Comment by jmike** [comment]
Cool, the code has been updated accordingly.

File: readme.md [hunk 2]
```diff
@@ -130,19 +130,19 @@ PR welcome for additional commonly needed types and docs improvements. Read the
...
+import type { Except } from "type-fest";
 
 type Foo = {
 	unicorn: string;
 	rainbow: boolean;
 };
 
-type FooWithoutRainbow = Except<Foo, 'rainbow'>;
+type FooWithoutRainbow = Except<Foo, "rainbow">;
 //=> {unicorn: string}
```
Comment (sindresorhus) [inline readme.md:146]: Don't make unrelated changes

Comment (jmike) [reply readme.md]: Apologies. This was unintentional.
